src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -47,6 +50,20 @@ public KafkaProducerMessageHandler(final KafkaProducerContext kafkaProducerConte
 		this.kafkaProducerContext = kafkaProducerContext;
 	}
 
+	/**
+	 * Enables the use of headers for determining the target topic and partition of outbound messages. By default it is
+	 * set to true, but it can be disabled when those values are produced by upstream components that read messages
+	 * from Kafka sources themselves.
+	 *
+	 * @see KafkaHeaders#TOPIC
+	 * @see KafkaHeaders#PARTITION_ID
+	 * @param enableHeaderRouting whether the topic and destination headers should be considered
```

Comment:
Interesting point. We have other components with similar `expression -> headers` logic.
And for such an issue to avoid the incoming headers we always recommend to use `<header-filter>`.

I can guess that this approach would be better because we bypass extra `Message<?>` creation after the `<header-filter>`. 
From other side maybe we should just always ask for those `expressions` rather than fail with similar issues...
I mean don't fallback to the headers at all.

CC @garyrussell